extensions/VisualEditor/wikia/modules/ve/ui/dialogs/ve.ui.WikiaInfoboxInsertDialog.js

```diff
@@ -0,0 +1,235 @@
+		} );
+	}
+};
+
+/**
+ * Prepare infobox template names
+ */
+ve.ui.WikiaInfoboxInsertDialog.prototype.loadInfoboxTemplates = function () {
+	this.getInfoboxTemplates().done(
+		this.showResults.bind( this )
+	);
+};
+
+/**
+ * Fetch infobox template names from API
+ */
+ve.ui.WikiaInfoboxInsertDialog.prototype.getInfoboxTemplates = function () {
+	var deferred;
+	if ( !this.gettingTemplateNames ) {
+		deferred = $.Deferred();
+		ve.init.target.constructor.static.apiRequest( {
+			action: 'query',
+			list: 'allinfoboxes'
+		} )
+			.done( function ( data ) {
+				deferred.resolve( data );
+			} )
+			.fail( function () {
+				// TODO: Add better error handling.
+				deferred.resolve( [] );
+			} );
+		this.gettingTemplateNames = deferred.promise();
+	}
+	return this.gettingTemplateNames;
+};
+
+/**
+ * Insert template
+ */
+ve.ui.WikiaInfoboxInsertDialog.prototype.insertInfoboxTemplate = function () {
+	ve.init.target.constructor.static.apiRequest({
+		action: 'visualeditor',
+		paction: 'parsefragment',
+		page: mw.config.get( 'wgRelevantPageName' ),
+		wikitext: this.transclusionModel.getWikitext()
+	}, { type: 'POST' } )
+		.done( this.onParseSuccess.bind( this ) )
+		.fail( function () {
+		// TODO: Implement some proper handling, at least tracking
+	}.bind( this ) );
+};
+
+/**
+ * Add the infobox template names to the dialog's SelectWidget
+ *
+ * @param {Object} data Response data from API
+ */
+ve.ui.WikiaInfoboxInsertDialog.prototype.showResults = function ( data ) {
+	var i,
+		items = [],
+		infoboxes = data.query ? data.query.allinfoboxes : [];
+
+	if ( infoboxes.length > 0 ) {
+		for ( i = 0; i < infoboxes.length; i++ ) {
+			items.push(
+				new OO.ui.DecoratedOptionWidget({
+					data: infoboxes[i].title,
+					label:  infoboxes[i].title
+				})
+			);
+		}
+		this.select.addItems( items );
+	}
+};
+
+/**
+ * Insert prepared linear model to surface.
+ */
+ve.ui.WikiaInfoboxInsertDialog.prototype.onParseSuccess = function () {
+	var type = 'wikiaInfoboxTransclusionBlock',
+		linmod = [
+			{
+				type: type,
+				attributes: {
+					mw: this.transclusionModel.getPlainObject()
+				}
+			},
+			{ type: '/' + type }
+		];
+
+	this.surface.getModel().getDocument().once( 'transact', this.onTransact.bind( this ) );
+	this.fragment = this.getFragment()
+		.collapseToEnd()
+		.setAutoSelect( true )
+		.insertContent( linmod );
+};
+
+/**
+ * Handle document model transaction
+ *
+ * Once the transclusionModel has inserted the transclusion, the new node in the surface will be selected.
+ * We can ask the commandRegistry for the command for the node and execute it.
+ */
+ve.ui.WikiaInfoboxInsertDialog.prototype.onTransact = function () {
+	ve.track( 'wikia', {
+		action: ve.track.actions.ADD,
+		label: 'dialog-infobox-insert'
+	} );
+	this.$frame.stopThrobbing();
+	this.close();
+
+	//setTimeout used for calling command for the node after current stack us cleared.
+	//getFocusedNode can be not known here yet
+	setTimeout( function () {
```

Comment:
Is there any other way to delay `execute()` method without using `setTimeout( function, 0)` pattern? If this is used across VE code maybe we could at least move it to separate helper method and later think how to do it delay execution without s`etTimeout`?